content/blog/2019-09-25-type_checker.md

```diff
@@ -0,0 +1,88 @@
+2. To ensure that the type checker works properly with labels, a first pass of the labels is designed. Through the first pass, all existing labels are saved and no duplicate ones are allowed to avoid conflicts. Though this almost doubles the type checking overhead, it further ensures the correctness of the program by static analysis.
+3. Keeping track of the line number and returning proper error message when encountered with an type checking error. When scanning through the code line by line after the first pass, a line number is maintained during type checking to help the programmer debug. In order to maintain the line number with blank lines and comments, we take in the original code snippet and ignore these lines while updating it during type checking.
+
+
+
+## Evaluation and results
+
+We wrote a set of benchmarking programs to test our implementation for various types of rules defined in the design section above. 
+
+The general test cases are classified into two sub-directories: [should-fail](https://github.com/tissue3/bril/tree/master/test/type-check/should-fail) and [should-pass](https://github.com/tissue3/bril/tree/master/test/type-check/should-pass). It has expected output named as "\*.out" corresponding to the input file "\*.bril". User can simply run all test cases in a directory by running turnt directory/\*.bril. For example, to run should-fail benchmark, one can just run turnt test/type-check/should-pass/\*.bril at the main directory of bril.
+
+Because should-pass cases are trivial, where we enumerate list existing operations, in the following table we only list tests cases that our type checker will report error message. The second column of the table aims to help one understand why an error would be reported by our type checker.
+
+| Instruction | Type Checking Rule |Testing Code Snippet  |
+| ----------- | :-----------: | :-------------: |
+| Conflict Definition       |   Variables cannot be redefined with a different type.   |      v: int = 5;<br/> v :bool = true;      |
+| Arithmetic         |          Adding an integer and a boolean type.           |                a: int = 4;<br/>b: bool = true;<br/>c: int = add a b;               |
+| Boolean         |      Cannot assign the output of boolean to an integer.      |    b: bool = true;<br/> a:int = not b;    |
+| Const       |    Cannot assign an integer const to a bool variable.    |            a: int =  const true;             |
+| Cond Branch |    Only takes bool variable as input (and 2 labels).     |     a: int = 1;<br/> br a here there;<br/> here: print a;<br/>there: jmp here; |
+
+
+Finally, though not required, we also implemented other checking passes as long as an input stream is parsable by bril2json, an interpret for bril. That includes type existence checking, number checking, argument existence checking, label existence checking and label repentance checking.
```

Comment:
> bril2json, an interpret for bril.

To be clear, `bril2json` is not an interpreter. It's just a parser.

Reply:
I'm not quite sure what this paragraph is saying. What's the difference between the original version of the type checker described above and this new one? Do they not work on the same input format?